src/workerd/jsg/jsg.h

```diff
@@ -1989,6 +2013,39 @@ class Lock {
   // Returns a random UUID for this isolate instance. This is largely intended for logging and
   // diagnostic purposes.
 
+  auto withinHandleScope(auto&& fn) {
+    // Runs the given function synchronously with a v8::HandleScope on the stack.
+    // If the fn returns a v8::Local<T> or v8::MaybeLocal<T> type, then
+    // v8::EscapableHandleScope is used ensuring that the v8::Local<T> return
+    // value is properly handled.
+    using Ret = decltype(fn());
+    if constexpr (isV8Local<Ret>()) {
+      v8::EscapableHandleScope scope(v8Isolate);
+      return scope.Escape(fn());
+    } else if constexpr (isV8MaybeLocal<Ret>()) {
+      v8::EscapableHandleScope scope(v8Isolate);
+      return scope.EscapeMaybe(fn());
+    } else {
+      v8::HandleScope scope(v8Isolate);
+      return fn();
+    }
+  }
+
+  class ContextScope {
```

Comment:
(no action) I wonder if we can eventually get rid of this by having the lock itself apply the context scope. I guess today we still have code outside JSG which needs to take an isolate lock without a context but maybe that will eventually go away?